pkg/runtime/stack_trace.go

```diff
@@ -0,0 +1,165 @@
+}
+
+type StackTrace interface {
+	// ProgramCounters copies a slice of program counters to the argument.
+	// The program counters which begins with the final caller and ends with the initial/root caller.
+	ProgramCounters(PCs) int
+
+	// Frames is an analog of runtime.CallerFrames.
+	Frames() FramesIterator
+
+	// String implements fmt.Stringer.
+	String() string
+
+	// Len returns the amount of frames in the stack trace
+	Len() int
+}
+
+// PCs is a goroutine execution frames stack trace.
+type PCs []PC
+
+var _ StackTrace = (PCs)(nil)
+
+// Len implements interface StackTrace.
+func (s PCs) Len() int {
+	return len(s)
+}
+
+// ProgramCounters implements interface StackTrace.
+func (s PCs) ProgramCounters(out PCs) int {
+	return copy(out, s)
+}
+
+// CallersFrames is an equivalent of standard runtime.CallerFrames.
+func (s PCs) Frames() FramesIterator {
+	pcs := make([]uintptr, len(s))
+	for idx, pc := range s {
+		pcs[idx] = uintptr(pc)
+	}
+	return runtime.CallersFrames(pcs)
+}
+
+// String implements fmt.Stringer.
+func (s PCs) String() string {
+	var result strings.Builder
+	frames := s.Frames()
+	if frames == nil {
+		return "<invalid stack trace>"
+	}
+	frameDepth := 1
+	for {
+		frame, ok := frames.Next()
+		result.WriteString(fmt.Sprintf("%d. %s:%d: %s\n", frameDepth, frame.File, frame.Line, frame.Function))
+		frameDepth++
+		if !ok {
+			break
+		}
+	}
+	return result.String()
+}
+
+// CallerStackTrace returns the StackTrace of the current Caller (in current goroutine).
+func CallerStackTrace(callerPCFilter PCFilter) PCs {
+	if callerPCFilter == nil {
+		callerPCFilter = DefaultCallerPCFilter
+	}
+
+	pcs := pcsPool.Get().(*[]uintptr)
+	defer pcsPool.Put(pcs)
+
+	startIdx := 0
+	n := runtime.Callers(1, *pcs)
+	for i := 0; i < n; i++ {
+		pc := (*pcs)[i]
+		if callerPCFilter(pc) {
+			startIdx = i
+			break
+		}
+	}
+
+	result := make(PCs, n-startIdx)
+	for idx, pc := range (*pcs)[startIdx:n] {
+		result[idx] = PC(pc)
+	}
+	return result
+}
+
+// Frames is a slice of runtime.Frame
+type Frames []runtime.Frame
+
+var _ StackTrace = (Frames)(nil)
+
+// Len implements interface StackTrace.
+func (s Frames) Len() int {
+	return len(s)
+}
+
+// ProgramCounters implements interface StackTrace.
+func (s Frames) ProgramCounters(out PCs) int {
+	endIdx := len(s)
+	if len(out) < endIdx {
+		endIdx = len(out)
+	}
+	for idx, frame := range s[:endIdx] {
+		out[idx] = PC(frame.PC)
+	}
+	return endIdx
+}
+
+// CallersFrames is an equivalent of standard runtime.CallerFrames.
+func (s Frames) Frames() FramesIterator {
+	if len(s) == 0 {
+		return nil
+	}
+	return newFramesIterator(s)
+}
+
+type framesIterator struct {
+	Frames      Frames
+	CurPosition int
+}
+
+func newFramesIterator(s Frames) *framesIterator {
+	return &framesIterator{Frames: s}
+}
+
+var _ FramesIterator = (*framesIterator)(nil)
+
+// Next implements FramesIterator.
+func (i *framesIterator) Next() (runtime.Frame, bool) {
+	frame := i.Frames[i.CurPosition]
```

Comment:
if _Next()_ is called multiple times in sequence and the bool is not checked, this line can panic. In this case, I am not sure what is the desired 

Perhaps, something like this might solve this.
```suggestion
        if i.CurPosition >= len(i) {
            return <last_element|nil>, false
        }
	frame := i.Frames[i.CurPosition]
```

Reply:
It was intentional (due to performance reasons), but I took a look at [the standard `(*Frames).Next`](https://pkg.go.dev/runtime#Frames.Next) description, and you right:

> If it has already returned all call frames, Next returns a zero Frame.

Fixed.